### ISSUE TITLE
calculation result is incorrect on metal backend

### ISSUE THREAD

**Issue by michael8090** [huggingface/candle#1914]
With the following code, which does a simple add on two tensors:
```rust
            let pred = pred.get(0).unwrap();
            let y = y.get(0).unwrap();

            let pred_loss = (&pred + &y).unwrap();

            println!(
                "pred:\n {} \n y: \n{} \n pre_loss: {}\n\n",
                pred,
                y,
                pred_loss
            );
```

I get the following output:

```text
pred:
 [-355.3564]
Tensor[[1], f32, metal:4294969932] 
 y: 
[-2879.0269]
Tensor[[1], f32, metal:4294969932] 
 pre_loss: [-355.3564]
Tensor[[1], f32, metal:4294969932]


pred:
 [-477.5577]
Tensor[[1], f32, metal:4294969932] 
 y: 
[3433.1504]
Tensor[[1], f32, metal:4294969932] 
 pre_loss: [-477.5577]
Tensor[[1], f32, metal:4294969932]


pred:
 [-623.6874]
Tensor[[1], f32, metal:4294969932] 
 y: 
[-3975.5510]
Tensor[[1], f32, metal:4294969932] 
 pre_loss: [-623.6874]
Tensor[[1], f32, metal:4294969932]
```

We can see here, the pred_loss is the same as pred. With Cpu backend, the result is right.

Environment info:

name = "candle-core"
version = "0.4.1"
name = "candle-nn"
version = "0.4.1"
name = "candle-metal-kernels"
version = "0.4.1"

Chip: Apple M1 Max

**Comment by LaurentMazare** [comment]
Could you provide some small standalone repro for this, ideally using the github version? I just tried some very basic one and everything works as expected (also we have quite a bunch of tests that are working well).
```rust
fn main() -> Result<()> {
    let device = Device::new_metal(0)?;
    let a = Tensor::new(&[1f32], &device)?;
    let b = Tensor::new(&[3f32], &device)?;
    let c = (&a + &b)?;
    println!("{a}\n{b}\n{c}");
    Ok(())
}
```

```bash
[1.]
Tensor[[1], f32, metal:4294968416]
[3.]
Tensor[[1], f32, metal:4294968416]
[4.]
Tensor[[1], f32, metal:4294968416]
```

My guess is that the problem would only trigger when the tensors are views over other tensors or reshaped/narrowed/..., I've eyeballed the code and haven't spotted anything weird so a repro would be very helpful.

**Comment by michael8090** [comment]
Sure, here is the repo: https://github.com/michael8090/test-candle-metal

 Please see the code at line 172 to see the issue. The MSE is not right.

**Comment by LaurentMazare** [comment]
That's odd, I just ran your code both on metal and cpu mutiple times and the result looked ok to me, e.g. printing `y`, `pred` and then the loss (and 0.0026 is appropriately the square of 0.0512).
Maybe you could make some more deterministic repro that prints more details about the inconsistency?
```
===== y:
[[0.]]
Tensor[[1, 1], f32, metal:4294968416]
===== pred:
[[0.0512]]
Tensor[[1, 1], f32, metal:4294968416]
===== pred_loss:
[0.0026]
Tensor[[], f32, metal:4294968416]
```

**Comment by michael8090** [comment]
In fact the code is deterministic. I just update the code, and now it only print 4 times. Everytime I run the code, it always prints the following:

```text
     Running `target/debug/test-candle-metal`
===== pred_loss: [0.0026]
Tensor[[], f32, metal:4294969932]
step: 0    pred_loss: 0.0026226344
pred:
[0.0512]
Tensor[[1], f32, metal:4294969932]
 y:
[0.]
Tensor[[1], f32, metal:4294969932] 


===== pred_loss: [0.6465]
Tensor[[], f32, metal:4294969932]
step: 1    pred_loss: 0.6464708
pred:
[-0.8040]
Tensor[[1], f32, metal:4294969932]
 y:
[499.7917]
Tensor[[1], f32, metal:4294969932] 


===== pred_loss: [0.1289]
Tensor[[], f32, metal:4294969932]
step: 2    pred_loss: 0.12894265
pred:
[-0.3591]
Tensor[[1], f32, metal:4294969932]
 y:
[998.3342]
Tensor[[1], f32, metal:4294969932] 


===== pred_loss: [0.0165]
Tensor[[], f32, metal:4294969932]
step: 3    pred_loss: 0.016508345
pred:
[0.1285]
Tensor[[1], f32, metal:4294969932]
 y:
[1494.3811]
Tensor[[1], f32, metal:4294969932] 
```

**Comment by michael8090** [comment]
I tried and the issue goes away, thanks. but is it the expected behavior?

**Comment by LaurentMazare** [comment]
With the current code it's expected as it's undefined what happens between different devices, that said I'll have a look at transforming this into a hard error so that it's easier to spot when this happens.

**Comment by LaurentMazare** [comment]
See #1932 for the added check, thanks again for reporting this.

**Comment by LaurentMazare** [comment]
Closing as hopefully fixed now.
